File: elixir/010/010.exs
defmodule Euler do
  def sieve_of_eratosthenes(limit) when is_number(limit) and limit > 1 do
    sieve_of_eratosthenes(2, round(:math.sqrt(limit)), [], :lists.seq(3, limit, 2))
  end

  defp sieve_of_eratosthenes(prime, limit, primes, integers) when prime > limit do
    :lists.reverse([prime | primes]) ++ integers
  end

  defp sieve_of_eratosthenes(prime, limit, primes, integers) do
    [new_prime | new_integers] = integers |> Enum.filter(fn x -> rem(x, prime) != 0 end)

    sieve_of_eratosthenes(new_prime, limit, [prime | primes], new_integers)
  end
end

IO.puts Euler.sieve_of_eratosthenes(2000000) |> Enum.reduce(0, &(&1 + &2))

File: clojure/010/010.clj
(defn sieve_of_eratosthenes
  ([limit]
    (sieve_of_eratosthenes 2 (Math/round (Math/sqrt limit)) '() (range 3 (+ limit 1) 2))
  )
  ([prime, limit, primes, integers]
    (if (> prime limit)
      (concat (reverse (concat (list prime) primes)) integers)
      (let [new_integers (filter #(not= (mod % prime) 0) integers)]
        (sieve_of_eratosthenes (first new_integers) limit (concat (list prime) primes) (rest new_integers))
      )
    )
  )
)

(println (apply + (sieve_of_eratosthenes 2000000)))
